source/common/matcher/matcher.h

```diff
@@ -324,6 +324,8 @@ class MatchTreeFactory : public OnMatchFactory<DataType> {
           on_match.action().typed_config(), server_factory_context_.messageValidationVisitor(),
           factory);
 
+      // TODO(taoxuy): try to pass message by moving and let the created action take ownership
+      // of the message if needed, which avoid copy.
```

Comment:
*message is not making a copy I think, it returns a reference to the underlying data.